src/oidcop/oidc/token.py

```diff
@@ -352,6 +363,249 @@ def post_parse_request(
 
         return request
 
+class TokenExchangeHelper(TokenEndpointHelper):
+    """Implements Token Exchange a.k.a. RFC8693"""
+
+    def __init__(self, endpoint, config=None):
+        TokenEndpointHelper.__init__(self, endpoint=endpoint, config=config)
+
+        if config is None:
+            self.config = {}
+        else:
+            self.config = config
+
+        if "token_exchange" in self.endpoint.kwargs:
+            self.config["token_exchange"] = self.endpoint.kwargs["token_exchange"]
+
+        if (
+            "token_exchange" in self.config
+            and "policy" not in self.config["token_exchange"]
+        ):
+            raise KeyError("No token echange policy found")
+
+
+
+    def post_parse_request(self, request, client_id="", **kwargs):
+        request = TokenExchangeRequest(**request.to_dict())
+
+        if "client_id" not in request:
+            request["client_id"] = client_id
+
+        _context = self.endpoint.server_get("endpoint_context")
+
+        try:
+            keyjar = _context.keyjar
+        except AttributeError:
+            keyjar = ""
+
+        try:
+            request.verify(keyjar=keyjar, opponent_id=client_id)
+        except (
+            MissingRequiredAttribute,
+            ValueError,
+            MissingRequiredValue,
+            JWKESTException,
+        ) as err:
+            return self.endpoint.error_cls(
+                error="invalid_request", error_description="%s" % err
+            )
+
+        _mngr = _context.session_manager
+        try:
+            _session_info = _mngr.get_session_info_by_token(
+                request["subject_token"], grant=True
+            )
+        except (KeyError, UnknownToken):
+            logger.error("Subject token invalid.")
+            return self.error_cls(
+                error="invalid_request",
+                error_description="Subject token invalid"
+            )
+
+        token = _mngr.find_token(_session_info["session_id"], request["subject_token"])
+
+        resp = self.enforce_policy(request, token)
+        if isinstance(resp, TokenErrorResponse):
+            return resp
+
+        if token.is_active() is False:
+            return self.error_cls(
+                error="invalid_request", error_description="Subject token inactive"
+            )
+        return resp
+
+    def enforce_policy(self, request, token):
+        _context = self.endpoint.server_get("endpoint_context")
+
+        if not isinstance(token, (AccessToken, RefreshToken)):
+            return self.error_cls(
+                error="invalid_request", error_description="Wrong token type"
+            )
```

Comment:
This should be removed, we check this with `subject_token_types_supported`

Reply:
`subject_token_types_supported` is a check for the integrity of the token exchange request, but shouldn't we check that the actual `token` is of the equivalent class? 

What about we add a check upon a allowed mapping like:
```python
subject_token_types_supported = {
            "urn:ietf:params:oauth:token-type:access_token": AccessToken,
            "urn:ietf:params:oauth:token-type:refresh_token": RefreshToken
        }

if (
    request["subject_token_type"] in subject_token_types_supported
    and (
        not isinstance(token,
                subject_token_types_supported[request["subject_token_type"]]
        )
    )
):
    return self.error_cls(
        error="invalid_request", error_description="Wrong token type"
    )
```

Reply:
LGTM